### PR TITLE
Detect Malicious Behaviors Through Audit Logs - blog post

### LINKED ISSUE #437: Analyze Kubernetes API Server Audit logs with Falco
/area blog

**What would you like to be added**:
Demonstrating how we can use Falco with Kubernetes API Server to detect intrusions that happened in Kubernetes API Server through the audit logs

**Why is this needed**:
We don't have any blog posts that related to describe how we can use Falco to detect intrusions that happened in Kubernetes API Server through the audit logs, so, I thought that it would be nice to have a blog post that explains everything related to that topic.

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind design

> /kind content

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area blog

**What this PR does / why we need it**:

This PR explains how users enable Kubernetes Audit Logs and detect malicious behaviors through Falco based on Audit Logs.

**Which issue(s) this PR fixes**:

Fixes #437 

**Special notes for your reviewer**:

### CODE REVIEW

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+* [Kubernetes Response Engine, Part 1 : Falcosidekick + Kubeless](https://falco.org/blog/falcosidekick-reponse-engine-part-1-kubeless/)
+* [Kubernetes Response Engine, Part 2 : Falcosidekick + OpenFaas](https://falco.org/blog/falcosidekick-reponse-engine-part-2-openfaas/)
+* [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative](https://falco.org/blog/falcosidekick-reponse-engine-part-3-knative/)
+* [Kubernetes Response Engine, Part 4: Falcosidekick + Tekton](https://falco.org/blog/falcosidekick-response-engine-part-4-tekton/)
```
Comment (Issif) [inline content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:15]: New blog post about Argo is missing

**Review by leogr — COMMENTED** [review]
Just found minor issues and typos, otherwise SMGT!

Btw, great blog, congrats :clap:

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+* When did it happen?
+* Who initiated it?
+* On what did it happen?
+
+To get more detail about them, you can refer to this [link](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/).
+
+In this guide, I'll explain these following topics:
+
+* How you can enable Audit log feature in Kubernetes.
+* Wow you can enable embedded web server available within the Falco project.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+Now, let's create the ConfigMap.
+
+```shell
+$ kubectl create configmap aws-creds \
+  --from-literal aws_access_key_id=AKES20LNOA
+configmap/aws-creds created
+```
+
+As soon as you created the ConfigMap, Falco will detect that from the audit logs, and send an alert for that to the Falcosidekick, after that Falcosidekick send it to the Web UI. So, you should see a screen similar to the following:
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+     Memory: 1.2G
+     CGroup: /system.slice/k3s.service
+...
+```
+
+Now we are ready to test everything.
+
+### Test
+
+There is a rule called `Create/Modify Configmap With Private Credentials` in Falco for detecting ConfigMaps that contains private credentials from `Kubernetes Audit Logs`, you can see the details of the rule from this [link](https://github.com/falcosecurity/falco/blob/master/rules/k8s_audit_rules.yaml#L185). In order to test this, we should create a ConfigMap that contains a private credetials, but before doing that let's access the `Falcosidekick Web UI` to see what will happen once we create the malicious ConfigMap.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+* k3sup v0.11.0
+* kubectl v1.21.1
+* helm v3.5.4+g1b5edb6
+
+## Tutorial
+
+### Provision local Kubernetes Cluster
+
+There are various ways to provision a local Kubernetes cluster such as, `KinD`, `k3s`, `k0s`, `Minikube` etc. We are going to
+use k3s in this walkthrough, we're also going to use some tooling to make installation process of `k3s` cluster easier and quicker such as `k3sup` and `multipass`.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+Kubernetes Audit Logs are the records that explain what’s happening inside of this brain but if we look at them more technically the official documentation says `Kubernetes Audit logs provide a security-relevant, chronological set of records documenting the sequence of actions in a cluster.`, so by collecting and analyzing them we can answer these following questions:
+
+* What happened?
+* When did it happen?
+* Who initiated it?
+* On what did it happen?
+
+To get more detail about them, you can refer to this [link](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/).
+
+In this guide, I'll explain these following topics:
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+
+* [Kubernetes Response Engine, Part 1 : Falcosidekick + Kubeless](https://falco.org/blog/falcosidekick-reponse-engine-part-1-kubeless/)
+* [Kubernetes Response Engine, Part 2 : Falcosidekick + OpenFaas](https://falco.org/blog/falcosidekick-reponse-engine-part-2-openfaas/)
+* [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative](https://falco.org/blog/falcosidekick-reponse-engine-part-3-knative/)
+* [Kubernetes Response Engine, Part 4: Falcosidekick + Tekton](https://falco.org/blog/falcosidekick-response-engine-part-4-tekton/)
+
+Without further ado, let's explain what `Audit Logs` are, and which useful informations that they have for us.
+
+We can assume that `Kubernetes API Server` is alike a brain of the cluster. We all communicate with the cluster through `Kubernetes API Server`, that means that all the requests hit `Kubernetes API Server` first. So, sometimes we might want to know that what’s happening inside of that brain. At this point, the `Kubernetes Audit Log` comes in to the picture.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+* [Kubernetes Response Engine, Part 1 : Falcosidekick + Kubeless](https://falco.org/blog/falcosidekick-reponse-engine-part-1-kubeless/)
+* [Kubernetes Response Engine, Part 2 : Falcosidekick + OpenFaas](https://falco.org/blog/falcosidekick-reponse-engine-part-2-openfaas/)
+* [Kubernetes Response Engine, Part 3 : Falcosidekick + Knative](https://falco.org/blog/falcosidekick-reponse-engine-part-3-knative/)
+* [Kubernetes Response Engine, Part 4: Falcosidekick + Tekton](https://falco.org/blog/falcosidekick-response-engine-part-4-tekton/)
+
+Without further ado, let's explain what `Audit Logs` are, and which useful informations that they have for us.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 9]
```diff
@@ -0,0 +1,377 @@
+---
+title: "Detect Malicious Behaviour on Kubernetes API Server through Audit Logs"
+date: 2021-05-22
+author: Batuhan Apaydın
+slug: detect-malicious-behaviour-on-kubernetes-api-server-through-audit-logs
+---
+
+## Introduction
+
+We might not know that Falco is not just for detecting malicious behavior that involves making Linux system calls, in addition to that, Falco v0.13.0 adds Kubernetes Audit Events to the list of supported event sources. That means that, once your cluster is configured with audit log enabled, you can send audit logs as an events to Falco, then, you can write falco rules that can read these events and detect malicious or other notable activity. After that you can send notifications about the malicious activities or set up a `Kubernetes Response Engine` to take an actions based on them. If you want to take a look at the rules which Falco provides for Kubernetes Audit logs, you can refer to this [link](https://falco.org/docs/event-sources/kubernetes-audit). Also, if you want to learn more about that `Kubernetes Response Engine` is, there is series of articles about how to create a Kubernetes response engine with Falco, Falcosidekick and a FaaS.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: done

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md (hunk 1 above)
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:15]: done

**Comment by developer-guy** [comment]
> Just found minor issues and typos, otherwise SMGT!
> 
> Btw, great blog, congrats 👏

Now, I fixed the issues and typos, would you mind to re-review the PR, thank you so much @leogr @Issif 🙏

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 10]
```diff
@@ -0,0 +1,377 @@
+---
+title: "Detect Malicious Behaviour on Kubernetes API Server through Audit Logs"
+date: 2021-05-22
+author: Batuhan Apaydın
+slug: detect-malicious-behaviour-on-kubernetes-api-server-through-audit-logs
+---
+
+## Introduction
+
+We might not know that Falco is not just for detecting malicious behavior that involves making Linux system calls, in addition to that, Falco v0.13.0 adds Kubernetes Audit Events to the list of supported event sources. That means that, once your cluster is configured with audit log enabled, you can send audit logs as an events to Falco, then, you can write Falco rules that can read these events and detect malicious or other notable activity. After that, you can send notifications about the malicious activities or set up a `Kubernetes Response Engine` to take any action based on them. If you want to take a look at the rules which Falco provides for Kubernetes Audit logs, you can refer to this [link](https://falco.org/docs/event-sources/kubernetes-audit). Also, if you want to learn more about that `Kubernetes Response Engine` is, there is series of articles about how to create a Kubernetes response engine with Falco, Falcosidekick and a FaaS.
```
Comment (leogr) [inline content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:10]:
```suggestion
We might not know that Falco is not just for detecting malicious behavior that involves making Linux system calls, in addition to that, Falco v0.13.0 adds Kubernetes Audit Events to the list of supported event sources. That means that, once your cluster is configured with audit log enabled, you can send audit logs as events to Falco, then, you can write Falco rules that can read these events and detect malicious or other notable activity. After that, you can send notifications about the malicious activities or set up a `Kubernetes Response Engine` to take any action based on them. If you want to take a look at the rules which Falco provides for Kubernetes Audit logs, you can refer to this [link](https://falco.org/docs/event-sources/kubernetes-audit). Also, if you want to learn more about that `Kubernetes Response Engine` is, there is series of articles about how to create a Kubernetes response engine with Falco, Falcosidekick and a FaaS.
```

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+* When did it happen?
+* Who initiated it?
+* On what did it happen?
+
+To get more detail about them, you can refer to this [link](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/).
+
+In this guide, I'll explain the following topics:
+
+* How you can enable Audit log feature in Kubernetes.
+* How you can enable embedded web server available within the Falco project.
```
Comment (leogr) [inline content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:34]:
```suggestion
* How you can enable the embedded web server available within the Falco project.
```

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+* multipass v1.6.2
+* k3sup v0.11.0
+* kubectl v1.21.1
+* helm v3.5.4+g1b5edb6
+
+## Tutorial
+
+### Provision local Kubernetes Cluster
+
+There are various ways to provision a local Kubernetes cluster such as, `KinD`, `k3s`, `k0s`, `Minikube` etc. We are going to use k3s in this walkthrough, we're also going to use some tooling to make installation process of `k3s` cluster easier and quicker such as `k3sup` and `multipass`.
```
Comment (leogr) [inline content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:50]:
```suggestion
There are various ways to provision a local Kubernetes cluster such as, `KinD`, `k3s`, `k0s`, `Minikube` etc. We are going to use k3s in this walkthrough, we're also going to use some tooling to make the installation process of `k3s` cluster easier and quicker such as `k3sup` and `multipass`.
```

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+
+```shell
+$ kubectl get nodes -o wide
+NAME         STATUS   ROLES         AGE   VERSION         INTERNAL-IP     EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
+k3s-leader   Ready    etcd,master   93s   v1.19.11+k3s1   192.168.64.26   <none>        Ubuntu 20.04.2 LTS   5.4.0-73-generic   containerd://1.4.4-k3s1
+```
+
+### Install Falco and Falcosidekick with Auditing Feature Enabled
+
+We'll use Helm to install the Falco, so, there is an option that we can enable or disable audit log feature called `auditLog.enabled`, once we set the value of the option as `true`, the embedded webserver will be started within the Falco to consume audit events from the port `8765` and behind `k8s-audit` endpoint.
```
Comment (leogr) [inline content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:159]:
```suggestion
We'll use Helm to install the Falco, so, there is an option that we can enable or disable the audit log feature called `auditLog.enabled`, once we set the value of the option as `true`, the embedded webserver will be started within the Falco to consume audit events from the port `8765` and behind `k8s-audit` endpoint.
```

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+     Memory: 1.2G
+     CGroup: /system.slice/k3s.service
+...
+```
+
+Now we are ready to test everything.
+
+### Test
+
+There is a rule called `Create/Modify Configmap With Private Credentials` in Falco for detecting ConfigMaps that contains private credentials from `Kubernetes Audit Logs`, you can see the details of the rule from this [link](https://github.com/falcosecurity/falco/blob/master/rules/k8s_audit_rules.yaml#L185). In order to test this, we should create a ConfigMap that contains private credetials, but before doing that, let's access the `Falcosidekick Web UI` to see what will happen once we create the malicious ConfigMap.
```
Comment (leogr) [inline content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md:355]:
```suggestion
There is a rule called `Create/Modify Configmap With Private Credentials` in Falco for detecting ConfigMaps that contains private credentials from `Kubernetes Audit Logs`, you can see the details of the rule from this [link](https://github.com/falcosecurity/falco/blob/master/rules/k8s_audit_rules.yaml#L185). In order to test this, we should create a ConfigMap that contains private credentials, but before doing that, let's access the `Falcosidekick Web UI` to see what will happen once we create the malicious ConfigMap.
```

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md (hunk 10 above)
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: fixed, thank you so much.

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md (hunk 12 above)
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: fixed, thank you so much.

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md (hunk 14 above)
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: fixed, thank you so much.

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md (hunk 13 above)
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: fixed, thank you so much.

File: content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md (hunk 11 above)
Comment (developer-guy) [reply content/en/blog/falco-detect-malicious-behavior-through-audit-logs.md]: fixed, thank you so much.

**Review by leogr — APPROVED** [review]
:rocket:

**Pull request merged**